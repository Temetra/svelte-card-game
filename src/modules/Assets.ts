export * from "@/modules/Assets/Audio";
export * from "@/modules/Assets/CardImages";
export * from "@/modules/Assets/LoadingStatus";
export * from "@/modules/Assets/Text";