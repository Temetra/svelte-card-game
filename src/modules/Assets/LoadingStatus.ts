import { writable } from "svelte/store";

export interface LoadingStatus { 
	name: string, 
	progress: number, 
	total: number 
}

export const loadingStatus = writable<LoadingStatus>(null);