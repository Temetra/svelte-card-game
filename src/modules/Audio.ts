import { checkResponse } from "./Fetching";

const AudioContext = window.AudioContext || window.webkitAudioContext;
const audioCtx = new AudioContext();
const sounds: Partial<Record<string, AudioBuffer>> = {};

export function prepareAudio() {
	return Promise.all([
		loadAudio("sounds/blip.mp3"),
		loadAudio("sounds/card-hard.mp3"),
		loadAudio("sounds/card-soft.mp3")
	])
	.then(values => {
		sounds.blip = values[0];
		sounds.card = values[1];
		sounds.slide = values[2];
	});
}

function loadAudio(filename: string): Promise<AudioBuffer> {
	return fetch(filename)
		.then(checkResponse)
		.then(response => response.arrayBuffer())
		.then(arrayBuffer => audioCtx.decodeAudioData(arrayBuffer));
}

export function playSound(name: string) {
	let buffer = sounds[name];

	if (buffer) {
		const trackSource = audioCtx.createBufferSource();
		trackSource.buffer = buffer;
		trackSource.connect(audioCtx.destination);
		trackSource.start();
	}
}
