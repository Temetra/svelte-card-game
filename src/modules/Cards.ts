export * from "@/modules/Cards/Card";
export * from "@/modules/Cards/CardState";
export * from "@/modules/Cards/GetCardName";
export * from "@/modules/Cards/Rank";
export * from "@/modules/Cards/Suit";
export * from "@/modules/Cards/Scoring";
