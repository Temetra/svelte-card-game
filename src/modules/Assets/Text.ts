import { checkResponse } from "@/modules/Fetching";

// Very basic implementation to get ball rolling

let text = {};

export async function prepareText(languageOverride: string = null) {
	// Get major language from override or client
	let lang = languageOverride || getLangFromClient();

	await fetch(`i18n/${lang}.json`)
		.then(checkResponse)
		.then(response => response.json())
		.then(json => text = json);
}

export function geti18nText(path: string, source = text): string {
	return path.split(".").reduce(objectReducer, source);
}

// Returns a property in an object for the given key
// Used in a reducer to get a nested property for a period-delimited path
function objectReducer(obj: any, key: string): any {
	return obj ? obj[key] : undefined;
}


// Default to en
function getLangFromClient(): string {
	return (typeof window !== 'undefined' && navigator.language) 
		? navigator.language.split("-")[0] 
		: "en";
}