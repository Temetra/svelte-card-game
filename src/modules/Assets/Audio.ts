import { loadingStatus } from "@/modules/Assets/LoadingStatus";
import { fetchFiles } from "@/modules/Fetching";

// Audio context
let audioCtx: AudioContext;

// In-memory cache for sounds
let sounds: Record<string, AudioBuffer>;

// Finds and plays sound. No-op until sounds are loaded.
export let playSound = async (name: string) => {};

// Returns promise to preload sounds into bank
export function prepareAudio() {
	const AudioContext = window.AudioContext || window.webkitAudioContext;

	if (AudioContext) {
		audioCtx = new AudioContext();
		
		const files = {
			"blip": "sounds/blip.mp3",
			"card": "sounds/card-hard.mp3",
			"slide": "sounds/card-soft.mp3"
		};
	
		// Return a promise to load images
		return fetchFiles(files, processResponse, updateProgress)
			.then(data => sounds = data)
			.then(() => playSound = playSoundInternal);
	}
}

// Get ArrayBuffer from response, convert to AudioBuffer
async function processResponse(response: Response): Promise<AudioBuffer> {
	return await response.arrayBuffer()
		.then(arrayBuffer => audioCtx.decodeAudioData(arrayBuffer));
}

// Update progress store
function updateProgress(progress: number, total: number) {
	loadingStatus.set({ name:"sounds", progress, total });
}

// Finds and plays sound
async function playSoundInternal(name: string) {
	let buffer = sounds[name];
	if (buffer) {
		const trackSource = audioCtx.createBufferSource();
		trackSource.buffer = buffer;
		trackSource.connect(audioCtx.destination);
		trackSource.start();
	}
}
