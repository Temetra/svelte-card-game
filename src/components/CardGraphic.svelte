<script lang="ts">
	import { createEventDispatcher } from "svelte";
	import type { StatefulCard } from "@/modules/Cards";
	import { Suit, Rank, CardState, getCardName } from "@/modules/Cards";
	import { getCardImage, getCardFace } from "@/modules/Assets";

	export let card: StatefulCard = { suit: Suit.Joker, rank:Rank.Ace, state:0 };

	// Dispatch card click events
	const dispatch = createEventDispatcher();
	const handleClick = () => dispatch("click", card);
	const handleEnter = () => dispatch("hover", { target:card, enter:true });
	const handleExit = () => dispatch("hover", { target:card, enter:false });
</script>

<style type="text/scss">
	section {
		position: relative;
		width: 100%;
		height: 100%;
		padding-top: 71.43%;
		transform-style: preserve-3d;
		transform: scale3d(1,1,1) translateY(0px) translateZ(0px) rotateX(0deg) rotateY(0deg);
		transition: transform 250ms;


		img {
			position: absolute;
			transform-style: preserve-3d;
			backface-visibility: hidden;
			top: 0;
			left: 0;
			bottom: 0;
			right: 0;

			&.texture {
				backface-visibility: visible;
			}

			&.back {
				transform: rotateY(180deg);
			}
		}

		&.deal {
			animation: deal-card 400ms forwards;
		}

		&.spin {
			animation-play-state: running;
			animation: spin-card 3000ms infinite;
		}

		&.hide {
			opacity: 0;
		}

		&.flipped {
			transform: scale3d(1,1,1) translateY(0px) translateZ(0px) rotateX(0deg) rotateY(180deg);
		}

		&.focused {
			transform: scale3d(1.25,1.25,1) translateY(50px) translateZ(100px) rotateX(-30deg) rotateY(0deg);
		}

		&.flipped.focused {
			transform: scale3d(1.25,1.25,1) translateY(50px) translateZ(100px) rotateX(-30deg) rotateY(180deg);
		}

		&.discard {
			img {
				backface-visibility: visible;

				&.face {
					opacity: 0;
				}

				&.back {
					transform: rotateY(0deg);
				}
			}

			animation: discard-card 400ms forwards;
		}
	}

	@keyframes spin-card {
		0% {
			transform: rotateY(0deg);
		}
		25% {
			transform: rotateY(360deg);
		}
		100% {
			transform: rotateY(360deg);
		}
	}

	@keyframes deal-card {
		from {
			opacity:0;
			transform: scale3d(1,1,1) translateY(-100px) translateZ(100px) rotateX(-30deg) rotateY(0deg);
		}
		to {
			transform: scale3d(1,1,1) translateY(0px) translateZ(0px) rotateX(0deg) rotateY(0deg);
		}
	}

	@keyframes discard-card {
		from {
			transform: scale3d(1,1,1) translateY(0px) translateZ(0px) rotateX(0deg) rotateY(180deg);
		}
		to {
			opacity: 0;
			transform: scale3d(1,1,1) translateY(-100px) translateZ(0px) rotateX(0deg) rotateY(180deg);
		}
	}
</style>

<section 
	class:flipped={(card.state & CardState.Flipped) != 0}
	class:focused={(card.state & CardState.Focused) != 0}
	class:deal={(card.state & CardState.Dealing) != 0}
	class:spin={(card.state & CardState.Spinning) != 0}
	class:hide={(card.state & CardState.Hidden) != 0}
	class:discard={(card.state & CardState.Discarding) != 0}
	on:click={handleClick}
	on:mouseover={handleEnter}
	on:mouseout={handleExit}
>
	<img src={getCardImage("CardTexture")} class="texture" alt="" />
	<img src={getCardFace(card)} class="face" alt={getCardName(card.suit, card.rank)} />
	<img src={getCardImage("CardBackgroundRed")} class="back" alt="" />
</section>