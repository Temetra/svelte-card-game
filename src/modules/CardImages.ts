import { Suit, Rank, Suits, Ranks } from "./Cards";
import { fetchFiles } from "./Fetching";
import { loading } from "@/stores/datastore";

// In-memory cache for card images
let images: Record<string, string>;

// Returns promise to preload images into bank
export function prepareCardGraphics() {
	// Create an array of images to load
	const files = {
		"CardBackground": "images/cards/CardBackground.svg",
		"Joker": "images/cards/Joker.svg"
	};
	
	// Iterate over card enums, adding a card for each combo
	for (const suit of Suits) {
		for (const rank of Ranks) {
			let name = `${Suit[suit]}${Rank[rank]}`;
			files[name] = `images/cards/${name}.svg`;
		}
	}

	// Return a promise to load images
	return fetchFiles(files, processResponse, updateProgress)
		.then(data => images = data);
}

// Get blob from response, convert to base64
async function processResponse(response: Response): Promise<string> {
	return await response.blob()
		.then(blob => new Promise<string>(resolve => {
			const reader = new FileReader();
			reader.addEventListener("error", () => resolve(null));
			reader.addEventListener("load", evt => resolve(evt.target.result as string));
			reader.readAsDataURL(blob);
		}));
}

// Update progress store
function updateProgress(progress: number, total: number) {
	loading.set({ name:"images", progress, total });
}

// Returns base64 image blob for card face background
export function getCardFace() : string {
	return images["CardBackground"];
}

// Returns base64 image blob for card face value
export function getCardValue(suit: Suit, rank: Rank) : string {
	switch (suit) {
		case Suit.Joker: return images["Joker"];
		default: return images[`${Suit[suit]}${Rank[rank]}`];
	}
}
