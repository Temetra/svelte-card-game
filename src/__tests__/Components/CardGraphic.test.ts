import "../Mocks/AudioContext";
import "../Mocks/Fetch";
import { StatefulCard, Suit, Rank, CardState } from "@/modules/Cards";
import { prepareCardGraphics } from "@/modules/Assets";
import CardGraphic from "@/components/CardGraphic.svelte";

beforeAll(async () => {
	await prepareCardGraphics();
});

beforeEach(() => {
	document.body.innerHTML = "";
});

test("CardGraphic", async () => {
	// Setup
	let card: StatefulCard = { suit: Suit.Hearts, rank:Rank.Queen, state:0 };

	// Render component
	let component = new CardGraphic({ target: document.body, props: { card } });
	expect(component).not.toBeNull();

	// Run tests
	let graphic = document.body.getElementsByTagName("section");
	expect(graphic.length).toBe(1);
	expect(graphic[0].classList.length).toBe(1);
	expect(graphic[0]).toMatchSnapshot();
});

test("flipped focused", async () => {
	// Setup
	let card: StatefulCard = { suit: Suit.Hearts, rank:Rank.Queen, state: CardState.Flipped | CardState.Focused };

	// Render component
	let component = new CardGraphic({ target: document.body, props: { card } });
	expect(component).not.toBeNull();

	// Run tests
	let graphic = document.body.getElementsByTagName("section");
	expect(graphic.length).toBe(1);
	expect(graphic[0].classList.length).toBe(3);
	expect(graphic[0].classList.contains("flipped")).toBeTruthy();
	expect(graphic[0].classList.contains("focused")).toBeTruthy();
});
