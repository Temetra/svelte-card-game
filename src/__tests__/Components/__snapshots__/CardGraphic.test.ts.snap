// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGraphic 1`] = `
<img
  alt="Queen of Hearts"
  class="face svelte-1lmz6nk"
  src="data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiIHN0YW5kYWxvbmU9Im5vIj8+DQo8IURPQ1RZUEUgc3ZnIFBVQkxJQyAiLS8vVzNDLy9EVEQgU1ZHIDEuMS8vRU4iICJodHRwOi8vd3d3LnczLm9yZy9HcmFwaGljcy9TVkcvMS4xL0RURC9zdmcxMS5kdGQiPg0KPHN2ZyB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiB2aWV3Qm94PSIwIDAgNTAwIDcwMCIgdmVyc2lvbj0iMS4xIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB4bWw6c3BhY2U9InByZXNlcnZlIiB4bWxuczpzZXJpZj0iaHR0cDovL3d3dy5zZXJpZi5jb20vIiBzdHlsZT0iZmlsbC1ydWxlOmV2ZW5vZGQ7Y2xpcC1ydWxlOmV2ZW5vZGQ7c3Ryb2tlLWxpbmVqb2luOnJvdW5kO3N0cm9rZS1taXRlcmxpbWl0OjI7Ij4NCiAgICA8cmVjdCBpZD0iSGVhcnRzUXVlZW4iIHg9IjAiIHk9IjAiIHdpZHRoPSI1MDAiIGhlaWdodD0iNzAwIiBzdHlsZT0iZmlsbDpub25lOyIvPg0KICAgIDxwYXRoIGQ9Ik00MTMuMDIzLDQxOS45NDVjMS4zNDcsNS4xMjcgMy4yNjksOS45MzkgNS43NjgsMTQuNDM3YzIuNDk4LDQuNDk3IDcuMzMyLDExLjQ4MiAxNC41MDIsMjAuOTU1YzUuMjU4LDYuOTUyIDguNDk1LDExLjM2MiA5LjcxMiwxMy4yMzFjMS45OTksMy4wNDIgMy40NDMsNS44MzMgNC4zMzQsOC4zNzVjMC44OTEsMi41NDIgMS4zMzYsNS4xMTcgMS4zMzYsNy43MjRjMCw0LjgyMyAtMS42MDgsOC44NjQgLTQuODIzLDEyLjEyM2MtMy4yMTUsMy4yNTkgLTcuMTkxLDQuODg5IC0xMS45MjgsNC44ODljLTQuNzc5LC0wIC04LjkyOSwtMS42OTUgLTEyLjQ0OSwtNS4wODRjLTIuNjUsLTIuNTIxIC00LjgwMSwtNi4yNzkgLTYuNDUyLC0xMS4yNzZjLTEuNDM0LDQuOTEgLTMuNDU1LDguNjQ3IC02LjA2MiwxMS4yMTFjLTMuNjA3LDMuNDc2IC03Ljc3OCw1LjIxNCAtMTIuNTE0LDUuMjE0Yy00LjY5MywtMCAtOC42NjksLTEuNjE5IC0xMS45MjgsLTQuODU2Yy0zLjI1OSwtMy4yMzcgLTQuODg4LC03LjA5NCAtNC44ODgsLTExLjU2OWMtMCwtMy45MTEgMC45NTUsLTcuOTg1IDIuODY3LC0xMi4yMjFjMS45MTIsLTQuMjM3IDUuNjA2LC05Ljc4OCAxMS4wODEsLTE2LjY1M2M3LjEyNiwtOC45OTUgMTIuMzE5LC0xNi4zODIgMTUuNTc3LC0yMi4xNjFjMi41NjQsLTQuNTYzIDQuNTE5LC05LjM0MiA1Ljg2NywtMTQuMzM5WiIgc3R5bGU9ImZpbGw6cmdiKDIxMCwwLDApO2ZpbGwtcnVsZTpub256ZXJvOyIvPg0KICAgIDxwYXRoIGQ9Ik00MjguNzcxLDU4My41ODZjLTUuNDI5LC0wLjYyMyAtMTAuODU3LC0zLjExNSAtMTYuMjg2LC03LjQ3NmMtMy4zODEsLTIuNjY5IC03LjY5NywtNi45NDEgLTEyLjk0OCwtMTIuODE0bDAsNTkuOTM1YzAsMy4yMDMgMS41MzUsNi4wOTYgNC42MDYsOC42NzZjMy4wNywyLjU4MSA2LjM4NSwzLjg3MSA5Ljk0NCwzLjg3MWMzLjU2LDAgNi44OTcsLTEuMjkgMTAuMDEyLC0zLjg3MWMzLjExNCwtMi41OCA0LjY3MiwtNS40NzMgNC42NzIsLTguNjc2bC0wLC0zOS42NDVabS0yMi42OTMsLTMzLjA3NGwxNy4yMiwxNy4wMjJjMS40MjQsMS40NDIgMy4yNDgsMi4zNDMgNS40NzMsMi43MDNsLTAsLTkuNzI4Yy0wLC0zLjE1MiAtMS41NTgsLTYuMDM0IC00LjY3MiwtOC42NDZjLTMuMTE1LC0yLjYxMiAtNi40NTIsLTMuOTE4IC0xMC4wMTIsLTMuOTE4Yy0yLjc1OCwtMCAtNS40MjgsMC44NTYgLTguMDA5LDIuNTY3Wm0tMjkuNSwtMTQuMThjNS4yNSwtMCA5LjAzMiwwLjIyMiAxMS4zNDYsMC42NjdjMi42NywwLjYyMyA1LjI1LDEuODY5IDcuNzQyLDMuNzM4YzUuNjA3LC00LjUzOSAxMS43NDcsLTYuODA4IDE4LjQyMSwtNi44MDhjNy4yMDksLTAgMTMuNjYsMi41OCAxOS4zNTYsNy43NDJjNS45NjIsNS4zMzkgOC45NDMsMTEuNTY5IDguOTQzLDE4LjY4OGwwLDkuODc4bDcuMzQyLC0wbC0wLDEyLjgxNWMtMC4yNjcsMC4wODkgLTIuNzE0LDAuMjY3IC03LjM0MiwwLjUzNGwwLDM5LjY0NWMwLDcuMjA4IC0yLjk4MSwxMy40ODIgLTguOTQzLDE4LjgyMWMtNS42OTYsNS4xNjIgLTEyLjE0Nyw3Ljc0MyAtMTkuMzU2LDcuNzQzYy03LjIwOCwtMCAtMTMuNzI3LC0yLjYyNiAtMTkuNTU1LC03Ljg3NmMtNS44MjksLTUuMjUxIC04Ljc0NCwtMTEuNDggLTguNzQ0LC0xOC42ODhsMCwtNjIuODcyYzAsLTMuMDI2IDAuNjIzLC02LjAwNyAxLjg2OSwtOC45NDRjLTEuMjQ2LC0wLjggLTIuNDAzLC0xLjMzNCAtMy40NzEsLTEuNjAxbC03LjYwOCwtMGwtMCwtMTMuNDgyWiIgc3R5bGU9ImZpbGw6cmdiKDIxMCwwLDApO2ZpbGwtcnVsZTpub256ZXJvOyIvPg0KICAgIDxwYXRoIGQ9Ik04Ni45NzgsMjgwLjA2MWMtMS4zNDcsLTUuMTI4IC0zLjI3LC05Ljk0IC01Ljc2OCwtMTQuNDM4Yy0yLjQ5OSwtNC40OTcgLTcuMzMzLC0xMS40ODIgLTE0LjUwMywtMjAuOTU1Yy01LjI1OCwtNi45NTMgLTguNDk1LC0xMS4zNjMgLTkuNzEyLC0xMy4yMzJjLTEuOTk4LC0zLjA0MSAtMy40NDMsLTUuODMzIC00LjMzNCwtOC4zNzVjLTAuODkxLC0yLjU0MiAtMS4zMzYsLTUuMTE3IC0xLjMzNiwtNy43MjRjLTAsLTQuODIzIDEuNjA4LC04Ljg2NSA0LjgyMywtMTIuMTI0YzMuMjE2LC0zLjI1OSA3LjE5MiwtNC44ODggMTEuOTI4LC00Ljg4OGM0Ljc4LC0wIDguOTMsMS42OTUgMTIuNDUsNS4wODRjMi42NSwyLjUyIDQuODAxLDYuMjc5IDYuNDUyLDExLjI3NmMxLjQzNCwtNC45MSAzLjQ1NSwtOC42NDcgNi4wNjIsLTExLjIxMWMzLjYwNywtMy40NzYgNy43NzgsLTUuMjE0IDEyLjUxNSwtNS4yMTRjNC42OTMsLTAgOC42NjksMS42MTggMTEuOTI4LDQuODU2YzMuMjU5LDMuMjM3IDQuODg4LDcuMDkzIDQuODg4LDExLjU2OWMwLDMuOTExIC0wLjk1Niw3Ljk4NSAtMi44NjgsMTIuMjIxYy0xLjkxMiw0LjIzNyAtNS42MDUsOS43ODggLTExLjA4LDE2LjY1NGMtNy4xMjcsOC45OTUgLTEyLjMxOSwxNi4zODIgLTE1LjU3OCwyMi4xNjFjLTIuNTY0LDQuNTYzIC00LjUyLDkuMzQzIC01Ljg2NywxNC4zNFoiIHN0eWxlPSJmaWxsOnJnYigyMTAsMCwwKTtmaWxsLXJ1bGU6bm9uemVybzsiLz4NCiAgICA8cGF0aCBkPSJNNzEuMjMsMTE2LjQxNmM1LjQyOCwwLjYyMyAxMC44NTcsMy4xMTUgMTYuMjg1LDcuNDc1YzMuMzgyLDIuNjcgNy42OTgsNi45NDIgMTIuOTQ5LDEyLjgxNWwtMCwtNTkuOTM2Yy0wLC0zLjIwNCAtMS41MzUsLTYuMDk2IC00LjYwNSwtOC42NzdjLTMuMDcxLC0yLjU4MSAtNi4zODYsLTMuODcxIC05Ljk0NSwtMy44NzFjLTMuNTYsLTAgLTYuODk3LDEuMjkgLTEwLjAxMiwzLjg3MWMtMy4xMTUsMi41ODEgLTQuNjcyLDUuNDczIC00LjY3Miw4LjY3N2wtMCwzOS42NDZabTIyLjY5MywzMy4wNzRsLTE3LjIyLC0xNy4wMjJjLTEuNDI0LC0xLjQ0MiAtMy4yNDgsLTIuMzQzIC01LjQ3MywtMi43MDNsLTAsOS43MjhjLTAsMy4xNTIgMS41NTcsNi4wMzQgNC42NzIsOC42NDdjMy4xMTUsMi42MTIgNi40NTIsMy45MTggMTAuMDEyLDMuOTE4YzIuNzU4LC0wIDUuNDI4LC0wLjg1NiA4LjAwOSwtMi41NjhabTI5LjUwMSwxNC4xODFjLTUuMjUxLDAgLTkuMDMzLC0wLjIyMiAtMTEuMzQ3LC0wLjY2N2MtMi42NjksLTAuNjIzIC01LjI1LC0xLjg2OSAtNy43NDIsLTMuNzM4Yy01LjYwNiw0LjUzOSAtMTEuNzQ3LDYuODA4IC0xOC40MjEsNi44MDhjLTcuMjA5LDAgLTEzLjY2MSwtMi41ODEgLTE5LjM1NiwtNy43NDJjLTUuOTYzLC01LjM0IC04Ljk0NCwtMTEuNTY5IC04Ljk0NCwtMTguNjg5bC0wLC05Ljg3OGwtNy4zNDIsLTBsLTAsLTEyLjgxNWMwLjI2NywtMC4wODkgMi43MTQsLTAuMjY3IDcuMzQyLC0wLjUzNGwtMCwtMzkuNjQ2Yy0wLC03LjIwOSAyLjk4MSwtMTMuNDgzIDguOTQ0LC0xOC44MjJjNS42OTUsLTUuMTYyIDEyLjE0NywtNy43NDMgMTkuMzU2LC03Ljc0M2M3LjIwOCwwIDEzLjcyNywyLjYyNiAxOS41NTYsNy44NzZjNS44MjksNS4yNTEgOC43NDMsMTEuNDggOC43NDMsMTguNjg5bDAsNjIuODczYzAsMy4wMjYgLTAuNjIzLDYuMDA3IC0xLjg2OSw4Ljk0NGMxLjI0NiwwLjgwMSAyLjQwMywxLjMzNSAzLjQ3MSwxLjYwMmw3LjYwOSwtMGwwLDEzLjQ4MloiIHN0eWxlPSJmaWxsOnJnYigyMTAsMCwwKTtmaWxsLXJ1bGU6bm9uemVybzsiLz4NCjwvc3ZnPg0K"
/>
`;
